### PR TITLE
MCO-564: Make NodeController aware of BuildController

### PR DESCRIPTION
**Pull request by cheesesashimi**
**- What I did**

This makes NodeController aware of BuildController. Specifically, it will
propagate the final OS image pullspec from the MachineConfigPool annotation
(`machineconfiguration.openshift.io/newestImageEquivalentConfig`) into the
currentImage and desiredImage annotations for each node in the target pool.

**- How to verify it**

I've written a small Go binary that can be used to help aid in these steps: https://github.com/openshift/machine-config-operator/pull/3852. It is not required, but does make the setup / teardown process of this much easier.

1. Bring up an OpenShift cluster.
2. Create a new MachineConfigPool or use an existing MachineConfigPool.
3. Opt a pool into layering by adding the `machineconfiguration.openshift.io/layering-enabled` label onto a MachineConfigPool.
4. Set `machineconfiguration.openshift.io/newestImageEquivalentConfig` to an arbitrary value, e.g., `registry.hostname.com/org/repo:tag-1`.
5. Set the `BuildSuccess` condition on the MachineConfigPool. Note: You may have to use my onclusterbuilds binary found here https://github.com/openshift/machine-config-operator/pull/3852 in order to easily do that: `$ ./onclustertesting set-build-status --pool=<poolname> --type BuildSuccess --status=true`
6. Add a node to a pool by setting the `node-role.kubernetes.io/<poolname>` label on the node.

At this point, you should see the MachineConfigPool transition into the `Updating` status. You should also see that the node you've added to the pool has the annotation `machineconfiguration.openshift.io/desiredImage: "registry.hostname.com/org/repo:tag-1"` present. Once the node gets the `machineconfiguration.openshift.io/currentImage: "registry.hostname.com/org/repo:tag-1"` annotation it will move onto the next node in the pool until no more nodes require an update. Adding the `currentImage` annotation will become the job of the MCD once https://github.com/openshift/machine-config-operator/pull/3848 lands. Until that PR lands, you will have to add the `currentImage` annotation to the node to simulate the MCD's behavior.

**- Description for the changelog**
Makes NodeController aware of BuildController

### CODE REVIEW

File: pkg/controller/node/node_controller.go [hunk 1]
```diff
@@ -955,18 +1092,19 @@ func (ctrl *Controller) setDesiredMachineConfigAnnotation(nodeName, currentConfi
 // getAllCandidateMachines returns all possible nodes which can be updated to the target config, along with a maximum
 // capacity.  It is the reponsibility of the caller to choose a subset of the nodes given the capacity.
 func getAllCandidateMachines(pool *mcfgv1.MachineConfigPool, nodesInPool []*corev1.Node, maxUnavailable int) ([]*corev1.Node, uint) {
-	targetConfig := pool.Spec.Configuration.Name
-	unavail := getUnavailableMachines(nodesInPool)
-	// If we're at capacity, there's nothing to do.
+	unavail := getUnavailableMachines(nodesInPool, pool)
+	klog.Infof("Unavailable node count: %d, maxUnavailable: %d", len(unavail), maxUnavailable)
```
Comment (yuqi-zhang) [inline pkg/controller/node/node_controller.go:1096]: nit: similar to above, slightly noisy, considering dropping verbosity of (mostly) debug useful logs

File: pkg/controller/node/node_controller.go [hunk 2]
```diff
@@ -816,9 +949,12 @@ func (ctrl *Controller) syncMachineConfigPool(key string) error {
...
 		// All the nodes that need to be upgraded should have `NodeUpdateInProgressTaint` so that they're less likely
 		// to be chosen during the scheduling cycle.
-		targetConfig := pool.Spec.Configuration.Name
+		// targetConfig := pool.Spec.Configuration.Name
 		hasInProgressTaint := checkIfNodeHasInProgressTaint(node)
-		if node.Annotations[daemonconsts.DesiredMachineConfigAnnotationKey] == targetConfig {
+
+		lns := ctrlcommon.NewLayeredNodeState(node)
+
+		if lns.IsDesiredEqualToPool(pool) {
```
Comment (yuqi-zhang) [inline pkg/controller/node/node_controller.go:957]: Just to understand, this check works for all pools layered or not, right?

File: pkg/controller/node/node_controller.go [hunk 3]
```diff
@@ -929,16 +1065,17 @@ func (ctrl *Controller) setDesiredMachineConfigAnnotation(nodeName, currentConfi
 			return err
 		}
 
-		newNode := oldNode.DeepCopy()
-		if newNode.Annotations == nil {
-			newNode.Annotations = map[string]string{}
-		}
-
-		if newNode.Annotations[daemonconsts.DesiredMachineConfigAnnotationKey] == currentConfig {
+		lns := ctrlcommon.NewLayeredNodeState(oldNode)
```
Comment (yuqi-zhang) [inline pkg/controller/node/node_controller.go:1068]: Same question as above, we're denoting "layered node state" as the new consolidated way to mark node state?

**Review by yuqi-zhang — COMMENTED** [review]
Some general questions/nits below, functionally seems fine to me.

/payload 4.14 nightly blocking

Although I think this wouldn't change behaviour for existing functionality, doing a payload test just to be safe since this is a large change.

One last ask: could you add more to the commit message as well? Maybe just port over the PR message so git history is a bit easier to parse with such a large change. Separate commits may also be helpful, but not really necessary

File: pkg/daemon/writer.go [hunk 4]
```diff
@@ -247,6 +248,23 @@ func implSetNodeAnnotations(client corev1client.NodeInterface, lister corev1list
 		for k, v := range m {
 			node.Annotations[k] = v
 		}
+
+		// HACK: Remove this.
```
[the remark this answers is not in the document]
Comment (cheesesashimi) [reply pkg/daemon/writer.go:252]: I didn't mean to include it in this PR. So I'll remove it completely 😄.

File: pkg/controller/node/node_controller.go (hunk 2 above)
Comment (cheesesashimi) [reply pkg/controller/node/node_controller.go:957]: Yes it does. It first determines if it should consider images or not as it determines equality based upon whether the pool is opted into layering and whether it has an image or not.

File: pkg/controller/node/node_controller.go (hunk 3 above)
Comment (cheesesashimi) [reply pkg/controller/node/node_controller.go:1068]: Yes. In the future, I'd like to port this to the MCD as well so that we have a singular place for these transitions and interrogations of the node object. That will also ensure that we can easily and rapidly test for edge-cases.

**Review by yuqi-zhang — APPROVED** [review]
Thanks for addressing the reviews, I think we're good to get this merged. There are some failures on the payload tests but... I can't really tell at what point they failed, and I don't think they are exposing any issues with this PR

/lgtm
